### PR TITLE
Versioned associations feature

### PR DESCRIPTION
**Pull request by charlie-wasp**
It's attempt to implement #19 

At the time it lacks some test cases, it covers only `belongs_to` and `has_many` associations.

And don't pay attention to a numerous commits, I'll squash it into one when code will be cleaned and fixed

### CODE REVIEW

File: spec/logidze/model_spec.rb [hunk 1]
```diff
@@ -280,4 +280,76 @@
...
+          }
+        )
+      end
+
+      it "returns association version, according to the owner" do
+        old_post = post.at(time(200))
+        expect(old_post.comments.first.content).to eql('My comment')
+
+        very_old_post = post.at(time(100))
+        expect(very_old_post.comments.length).to eql(0)
```
Comment (charlie-wasp) [inline spec/logidze/model_spec.rb]: Note the `length` method usage instead of `size`. It corresponds to the problem I described in the issue #19.
Unlike `size` `length` method use `load_target`, so we can get, what we want, here

**Comment by charlie-wasp** [comment]
Oops, travis is not satisfied with my code, I'll get to it tomorrow

**Comment by charlie-wasp** [comment]
Oops, travis is not satisfied with my code, I'll get to it tomorrow

File: lib/logidze/model.rb [hunk 2]
```diff
@@ -50,6 +50,9 @@ def at(ts)
 
       object_at = dup
       object_at.apply_diff(version, log_data.changes_to(version: version))
+      object_at.id = id
```
Comment (palkan) [inline lib/logidze/model.rb]: Looks like that was a bug. Can you, please, add specs for this and, maybe, it would be better to extract this fix into a separate PR to release it ASAP?

File: lib/logidze/model.rb [hunk 3]
```diff
@@ -122,6 +135,14 @@ def apply_diff(version, diff)
       self
     end
 
+    def in_the_past?
```
Comment (palkan) [inline lib/logidze/model.rb]: Let's prefix this method to avoid possible collisions, smth like `logidze_past?`.

File: lib/logidze/versioned_association.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+      return target if inversed
+
+      time = owner.log_data.current_version.time
+
+      if target.is_a? Array
+        target.map! do |object|
+          object unless has_logidze? object
+
+          object.at(time)
```
Comment (palkan) [inline lib/logidze/versioned_association.rb]: We can use `at!` here too to avoid `#dup` calls.

File: lib/logidze/versioned_association.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+module Logidze
+  module VersionedAssociation
+    def load_target
+      target = super
+
+      return target if inversed
```
Comment (palkan) [inline lib/logidze/versioned_association.rb]: Can you explain this line?

Also, specs are needed.

File: lib/logidze/versioned_association.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+          object.at(time)
+        end.compact!
+      elsif has_logidze? target
+        target.at!(time)
+      end
+
+      target
+    end
+
+    def has_logidze?(object)
```
Comment (palkan) [inline lib/logidze/versioned_association.rb]: Let's add `has_logidze?` class method in `has_logidze.rb` to simplify the check.

File: lib/logidze/model.rb (hunk 2 above)
Comment (charlie-wasp) [inline lib/logidze/model.rb]: If we consider, that results of `at` invocation couldn't access their association, as a bug, of course, I can extract this particular change

File: lib/logidze/versioned_association.rb (hunk 5 above)
Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: I'll try to explain it with the example:

when we call `post.comments`, each comment in this association will have its `post` association initialized as inversed. So I thought, that we don't want to perform our operations in this case. I'll check this more closely during specs writing

File: lib/logidze/model.rb (hunk 2 above)
Comment (charlie-wasp) [inline lib/logidze/model.rb]: @palkan I have opened #25

File: lib/logidze/versioned_association.rb (hunk 5 above)
Comment (palkan) [inline lib/logidze/versioned_association.rb]: Yep. But this `post` object would exactly the object we call association on, so it will be at the _correct_ time. But, nevertheless, we need specs.

**Comment by charlie-wasp** [comment]
@palkan want to notice, that I introduced a new model called `Article` in order to fix specs. It's basically the same as the `Post`, but with the `log_data` column by default. I realized, that `Post` doesn't have it at the start by design, because this model is used to test generators. So for not messing with that, I created a new model. Is it okay?

**Comment by palkan** [comment]
@charlie-wasp Yep, that's good.

File: lib/logidze/model.rb [hunk 7]
```diff
@@ -38,11 +38,15 @@ def without_logging(&block)
...
     TIME_FACTOR = 1_000
 
+    attr_reader :logidze_requested_ts
+
     # Return a dirty copy of record at specified time
     # If time is less then the first version, then return nil.
     # If time is greater then the last version, then return self.
     def at(ts)
       ts = parse_time(ts)
+      @logidze_requested_ts = ts
```
Comment (charlie-wasp) [inline lib/logidze/model.rb]: @palkan here's how I covered the case, when `at` doesn't change the object. Can we do better? I am not particularly happy with `attr_reader`

File: lib/logidze/versioned_association.rb (hunk 4 above)
Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: It's strange, but when I use `at`, test case with empty comments failed. Somehow `super` return `[false]` instead of `[]`, so `has_logidze?` fails. I guess, somehow `at!` messes up something inside. I can investigate it, if it's important

File: lib/logidze/model.rb [hunk 8]
```diff
@@ -117,6 +121,18 @@ def switch_to!(version)
...
     end
 
+    def association(name)
+      association = super
+
+      if logidze_past?
+        association.singleton_class.prepend Logidze::VersionedAssociation
+      end
+
+      association.reload if @logidze_requested_ts
```
Comment (charlie-wasp) [inline lib/logidze/model.rb]: It seems, that with this we can use `.size`!

File: lib/logidze/has_logidze.rb [hunk 9]
```diff
@@ -13,6 +13,10 @@ module ClassMethods # :nodoc:
       def has_logidze
         include Logidze::Model
       end
+
+      def has_logidze?
+        included_modules.include? Logidze::Model
```
Comment (charlie-wasp) [inline lib/logidze/has_logidze.rb]: Is it what your suggestion was about?

File: lib/logidze/versioned_association.rb (hunk 5 above)
Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: I made a spec for inversed association. Correct me, if I wrong: in case of `belongs_to` inversed `has_many` is not populated anyway, so we don't care about it, right? E.g. in pseudo-code
```
user = article.user
user.articles.first == article # > false
```

File: lib/logidze/has_logidze.rb (hunk 9 above)
Comment (palkan) [inline lib/logidze/has_logidze.rb]: No. I suggested to add:

```ruby
def has_logidze?
  true
end
```

to `Logidze::Model::ClassMethods`.

File: lib/logidze/model.rb (hunk 8 above)
Comment (palkan) [inline lib/logidze/model.rb]: Hm, but we don't want to reload the association on every call. So we have to add another variable to track associations reload (btw, we should track each association separately) to avoid double-reload:

```ruby
post = Post.find(id)

# first load, without versions
post.comments

post.at!(2.days.ago)

# here we should reload (it would be better to reset, not reload, btw)
post.comments.first

# no reset/reload here
post.comments.second

# with fresh object
post = Post.find(id).at(2.days.ago)

# no reset/reload, 'cause association hasn't been loaded before
post.comments
```

File: lib/logidze/versioned_association.rb (hunk 5 above)
Comment (palkan) [inline lib/logidze/versioned_association.rb]: Yep

File: lib/logidze/has_logidze.rb (hunk 9 above)
Comment (charlie-wasp) [inline lib/logidze/has_logidze.rb]: Mmm, I thought about that, but classes without `has_logidze` wouldn't respond to this method. So we would have to check it beforehand, which not very convenient. Am I missing something?

File: lib/logidze/model.rb (hunk 8 above)
Comment (charlie-wasp) [inline lib/logidze/model.rb]: Yes, it was pretty brutal :) I used `reload` to invoke `load_target` in case of already loaded before `at` invocation `belongs_to` association:
```
article.user
old_article = article.at(...)
old_article.user # here load_target will not be called, so we won't have versioned user
```

But I can work around this issue another way, without `reload` or `reset` at all. I pushed commit recently with these changes.

But it seems, that there is no clean way to achieve `size` method working, so I got back to `length` so far

File: lib/logidze/versioned_association.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+require 'active_support'
+
+module Logidze
+  module VersionedAssociation
+    module SingularAssociation
+      def reader(force_reload = false)
```
Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: It turned out, in case of singular association we can use `reader` method to version instead of `load_target`. It will save us from the problem of already loaded associations, and multiple invocation won't make us suffer, because logidze won't do anything on the similar `at` calls

File: lib/logidze/versioned_association.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        super_result = super
+        return super_result unless owner.logidze_requested_ts
+
+        logidze_modify_target(target, owner.logidze_requested_ts)
+      end
+    end
+
+    module CollectionAssociation
+      def load_target
```
Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: In case of collection association we use `load_target`, because it will be used by `CollectionProxy` object

File: lib/logidze/has_logidze.rb (hunk 9 above)
Comment (palkan) [inline lib/logidze/has_logidze.rb]: Yep, you're right. Than we can check for `has_logidze?` method itself too (i.e. `respond_to?(:has_logidze?)`. It's still a little bit faster than checking included modules.

File: lib/logidze/model.rb [hunk 12]
```diff
@@ -125,6 +145,14 @@ def apply_diff(version, diff)
       self
     end
 
+    def logidze_past?
+      return false unless log_data
+
+      time = log_data.current_version.time
```
Comment (palkan) [inline lib/logidze/model.rb]: We should use `logidze_requested_ts` here

File: lib/logidze/model.rb (hunk 7 above)
Comment (palkan) [inline lib/logidze/model.rb]: I'd prefer to store this value in `log_data` object.

File: lib/logidze/versioned_association.rb (hunk 10 above)
Comment (palkan) [inline lib/logidze/versioned_association.rb]: But we can still use `load_target` for single association too, can't we? I would prefer to stick with as less monkey-patched method as possible. Thus, if we can handle all cases within `load_target`, then let's do that.

Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: Yes, absolutely, but with the `load_target` only we have to deal with the reset/reload problem

Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: @palkan I've got an idea. We override only `load_path` and `association`, as it was before my last commit, but in `association` we add something like this:

```
if association.is_a?(ActiveRecord::Associations::SingularAssociation) &&
   association.loaded? &&
   !association.inversed
  # we can extract functionality from load_target and use it here
  association.target.at! @logidze_requested_ts
end
```

What do you think about it? The same level of monkey-patching and no reload/reset confusion :)

Comment (palkan) [inline lib/logidze/versioned_association.rb]: And how to handle association loading? I.e. `post.at(...).user`?
How to prevent double-`at!`?

I have another idea: we can use [`stale_target?`](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/association.rb#L75) to handle reload. Just:

```ruby
def stale_target?
  logidze_state? || super
end

def logidze_stale?
  # we can store logidze_requested_ts in the association too
  loaded? && !inversed? && (owner.logidze_requested_ts != logidze_requested_ts)
end
```

That should work for all kind of associations.

But I found one more problem with collection associations: we have to think about `empty?`, `any?`, `blank?` methods (which use SQL if target has not been loaded) and also `update_all`, `delete_all`.
For presence-like checks we can force load target; and I think we should raise an error when a user tries to use `update_all` / `delete_all` on the past version.
Potential caveat:

```ruby
post = Post.find(...)
post.comments.size #=> 2

# rollback and persist
post.undo!

# no raise
post.comments.update_all(...)
```

Too complicated. We need more specs, first of all)

UPD: one more issue: `post.comment_ids`.

Maybe it would be easier to write our own `CollectionAssociation` that mimics the AR one and make it readonly, than to monkey-patch everything.

Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: That's a lot of information to process :)

1. >And how to handle association loading? I.e. post.at(...).user?
How to prevent double-at!?
All double `at!` wouldn't affect association, because there will be the same timestamp, so `at!` will just return self without any processing. Or I miss some scenario?
2. > we can use stale_target?
What advantage does it have over explicit `reload` or `reset` call?
3. > I found one more problem with collection associations
Got it, I will start with specs for all these methods

Issue gone wild, didn't it? :)

Comment (palkan) [inline lib/logidze/versioned_association.rb]: > Issue gone wild, didn't it? :)

Oh, yeah)

`stale_target?` is a standard AR mechanism to check, whether we need reload or not, and it's used all over the associations codebase. Thus we're unlikely to miss something. And we should not care about `#reload` ourselves. I hope so, at least.

> All double at! wouldn't affect association ...

Ok.

**Comment by charlie-wasp** [comment]
Added usage of `stale_target?` method

**Comment by charlie-wasp** [comment]
@palkan I added specs and implementation of presence-like methods. You said
> For presence-like checks we can force load target

Did you mean, that we can basically do something like this?
```
def empty?
  reload
  super
end
```

**Comment by charlie-wasp** [comment]
@palkan what do you think about my last commit? And what have we got on our plate? I feel kind of lost :) What we should do about `update_all`? Also, we should take care in a similar way about `item_ids=` method, I suppose.

File: lib/logidze/versioned_association.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    def load_target
+      target = super
+
+      return target if inversed
+
+      time = owner.logidze_requested_ts
+
+      if target.is_a? Array
+        target.map! do |object|
+          object unless object.class.has_logidze?
```
Comment (palkan) [inline lib/logidze/versioned_association.rb]: Why do we check for logidze here? We've already checked that in `#association` method.

**Comment by palkan** [comment]
@charlie-wasp Implementing `update_all` for versioned associations would be a huge pain)
I think, we can start with read-only features (that should be the most popular case).

File: lib/logidze/versioned_association.rb (hunk 13 above)
Comment (charlie-wasp) [inline lib/logidze/versioned_association.rb]: Not actually, we don't check, whether target class has logidze or not in the `#association` method :) And now I think, that it would be better to check it there really

Should we also add specs for the case of associations without logidze enabled?

**Comment by charlie-wasp** [comment]
@palkan so we leave `update_all` and the similar alone or we should implement raising exception on their invocation?

**Comment by palkan** [comment]
> so we leave update_all and the similar alone or we should implement raising exception on their invocation?

Just ignore them. And we should specify in the Readme the list of **supported** methods and a note telling that other methods may not work with versions.

**Comment by charlie-wasp** [comment]
@palkan good. Now can you tell, what else should I do, so we can merge this particular PR? As far as I see, only place for storing `logidze_requested_ts` is not fixed yet

File: lib/logidze/model.rb [hunk 14]
```diff
@@ -123,6 +135,24 @@ def switch_to!(version, append: Logidze.append_on_undo)
       end
     end
 
+    def association(name)
+      association = super
+
+      unless logidze_past? && association.klass.respond_to?(:has_logidze?)
```
Comment (palkan) [inline lib/logidze/model.rb]: We should also check whether `VersionedAssociation` has been already prepended

File: lib/logidze/model.rb [hunk 15]
```diff
@@ -131,6 +161,14 @@ def apply_diff(version, diff)
       self
     end
 
+    def logidze_past?
+      return false unless @logidze_requested_ts
+
+      time = @logidze_requested_ts
```
Comment (palkan) [inline lib/logidze/model.rb]: We don't need this local var at all

File: lib/logidze/versioned_association.rb [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    def logidze_stale?
+      return false if !loaded? || inversed
+
+      unless target.is_a?(Array)
+        return owner.logidze_requested_ts != target.logidze_requested_ts
+      end
+
+      return false if target.empty?
+
+      owner.logidze_requested_ts != target.first.logidze_requested_ts
```
Comment (palkan) [inline lib/logidze/versioned_association.rb]: There can be a case when we revert one of the target records into a previous state:

```ruby
post = post.at(122)
post.comments.size #=> 2
post.comments.first.at!(100)
post.at!(100)

# this version is for 122 time
post.comments.second
```

So, it's better to use `target.any? { ... }` here.

File: lib/logidze/versioned_association.rb [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    module CollectionAssociation
+      def ids_reader
+        reload
```
Comment (palkan) [inline lib/logidze/versioned_association.rb]: Shouldn't we check whether association has been loaded? If it was, then we don't need to reload, do we? The same goes for `empty?`.

File: spec/dummy/app/models/article.rb [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+  belongs_to :user
+  has_many :comments
+
+  # def reload
+  #   self.errored = nil
+  #   super
+  # end
+
+  # def is_errored
```
Comment (palkan) [inline spec/dummy/app/models/article.rb]: Cleanup, please

**Comment by charlie-wasp** [comment]
@palkan are there any more changes to be done?

**Comment by palkan** [comment]
@charlie-wasp 

Final stuff:

- Great Readme entry with all the caveats explained (or maybe event Wiki article)

- Change log entry

- It would be better to rename the PR, it's confusing

As it turned out this feature is rather complex and looks error-prone. What do you think about making it optional and disabled by default? And to enable a user must run smth like `Logidze.enable_associations_versioning!`

**Comment by charlie-wasp** [comment]
@palkan

**Comment by palkan** [comment]
@charlie-wasp 
> At the moment Wiki seems to be disabled

Check again, please. I've added you to collaborators.

**Comment by charlie-wasp** [comment]
@palkan all set, can edit wiki, thank you!

**Comment by charlie-wasp** [comment]
@palkan I wrote a wiki page, and the README and CHANGELOG entries. Also, I made this feature optional and disabled by default

**Comment by charlie-wasp** [comment]
@palkan hello, any updates on this?

**Comment by palkan** [comment]
Great work! Thanks!

**Comment by charlie-wasp** [comment]
@palkan thank you for the opportunity to contribute and your patience, it was a long run :+1:
